### ISSUE TITLE
Error in the execution of a confluent-kafka-python producer in a Confluent Docker container for Kafka

### ISSUE THREAD

**Issue by garrisa** [confluentinc/confluent-kafka-python#393]
Description
===========
I'm trying to run a test of a producer example python file in a Confluent Docker container for Kafka, but doesn't work properly.

Im running a Confluence Docker kafka container an Zookeeper in my Mac OS:
~ dkpsa
CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS                     PORTS                          NAMES
a2f758013289        confluentinc/cp-kafka:4.1.0       "/etc/confluent/dock…"   29 hours ago        Exited (255) 3 hours ago   9092/tcp                       kafka
1c38f08ba1aa        confluentinc/cp-zookeeper:4.1.0   "/etc/confluent/dock…"   29 hours ago        Exited (255) 3 hours ago   2181/tcp, 2888/tcp, 3888/tcp   zookeeper

I installed the confluent-kafka-python package in the container

docker run \
  --net=confluent \
  --rm \
-it -v '/Volumes/G-DRIVE mobile SSD R-Series/Dockers/Users':/ConfluenceKafkaClients \
confluentinc/cp-kafka:4.1.0 \
pip install confluent-kafka

I execute the producer example in the container from a local folder in the container:

docker run \
  --net=confluent \
  --rm \
-it -v '/Volumes/G-DRIVE mobile SSD R-Series/Dockers/Users':/ConfluenceKafkaClients \
confluentinc/cp-kafka:4.1.0 \
python /ConfluenceKafkaClients/producer_example.py

but and error appear:

Traceback (most recent call last):
  File "/ConfluenceKafkaClients/producer_example.py", line 8, in <module>
    from confluent_kafka import Producer
ImportError: No module named confluent_kafka

The container doesn´t find the module confluent_kafka,  if I list the modules installed in the containers,
confluent_kafka doesn´t appear:

➜  ConfluenceKafkaClients docker run \
  --net=confluent \
  --rm \
-it -v '/Volumes/G-DRIVE mobile SSD R-Series/Dockers/Users':/ConfluenceKafkaClients \
confluentinc/cp-kafka:4.1.0 \
pip freeze
backports.ssl-match-hostname==3.5.0.1
cached-property==1.4.0
confluent-docker-utils==0.0.20
docker==2.5.1
docker-compose==1.16.1
docker-pycreds==0.2.2
dockerpty==0.4.1
docopt==0.6.2
enum34==1.1.6
functools32==3.2.3.post2
ipaddress==1.0.19
Jinja2==2.9.6
jsonschema==2.6.0
MarkupSafe==1.0
PyYAML==3.12
requests==2.11.1
six==1.11.0
texttable==0.9.1
websocket-client==0.47.0
You are using pip version 8.1.2, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

 in some way the container doesn't install the pip confluent_kafka package because it doesn't recognise the module.

How can I run the producer test in this environment? I'm doing wrong?

How to reproduce
================
Here, how run the confluent Zookeeper and Kafka:
~ docker run -d \
    --net=confluent \
    --name=zookeeper \
    -e ZOOKEEPER_CLIENT_PORT=2181 \
    confluentinc/cp-zookeeper:4.1.0

➜  ~ docker run \
  --net=confluent \
  --rm \
  confluentinc/cp-kafka:4.1.0 \
> python pro
➜  ~ dkpsa
CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS                     PORTS                          NAMES
a2f758013289        confluentinc/cp-kafka:4.1.0       "/etc/confluent/dock…"   29 hours ago        Exited (255) 3 hours ago   9092/tcp                       kafka
1c38f08ba1aa        confluentinc/cp-zookeeper:4.1.0   "/etc/confluent/dock…"   29 hours ago        Exited (255) 3 hours ago   2181/tcp, 2888/tcp, 3888/tcp   zookeeper

Checklist
=========
Please provide the following information:

 - [ ] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`):

librdkafka 0.11.4
confluent-kafka-0.11.4
 - [ ] Apache Kafka broker version:
https://hub.docker.com/r/confluentinc/cp-kafka/
Two docker containers with:
confluentinc/cp-kafka:4.1.0
confluentinc/cp-zookeeper:4.1.0

 - [ ] Client configuration: `{...}`

producer_example.py : 
_from confluent_kafka import Producer
 p = Producer({'bootstrap.servers': 'localhost:9092'})
p.produce('mytopic', key='hello', value='world')
p.flush(30)_
 - [ ] Operating system:
Local OS : macOS HighSierra 10.13.4
 - [ ] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

Traceback (most recent call last):
  File "/ConfluenceKafkaClients/producer_example.py", line 8, in <module>
    from confluent_kafka import Producer
ImportError: No module named confluent_kafka

The container doesn´t find the module confluent_kafka,  if I list the modules installed in the containers,
confluent_kafka doesn´t appear.

**Comment by edenhill** [comment]
```
docker run 
--net=confluent 
--rm 
-it -v '/Volumes/G-DRIVE mobile SSD R-Series/Dockers/Users':/ConfluenceKafkaClients 
confluentinc/cp-kafka:4.1.0 
pip install confluent-kafka
```

With --rm you are removing the container, including the freshly installed confluent-kafka module, when installation is complete.
Suggest you run it without --rm

You might also need to perform the next `docker run` on the docker image created in the first step

**Comment by garrisa** [comment]
Thank for your quick answer but it seems that this is not the problem, I didn’t without —rm and 

ConfluenceKafkaClients dkpsa
CONTAINER ID        IMAGE                             COMMAND                  CREATED             STATUS                      PORTS                          NAMES
a2f758013289        confluentinc/cp-kafka:4.1.0       "/etc/confluent/dock…"   46 hours ago        Exited (255) 20 hours ago   9092/tcp                       kafka
1c38f08ba1aa        confluentinc/cp-zookeeper:4.1.0   "/etc/confluent/dock…"   46 hours ago        Exited (255) 20 hours ago   2181/tcp, 2888/tcp, 3888/tcp   zookeeper

➜  ConfluenceKafkaClients docker run \
  --net=confluent \
> confluentinc/cp-kafka:4.1.0 \
pip install confluent-kafka
Collecting confluent-kafka
  Downloading https://files.pythonhosted.org/packages/5b/23/8041cff8372a1c34f8c9acc51c41a12399af2563f69a166d8a395076309a/confluent_kafka-0.11.4-cp27-cp27mu-manylinux1_x86_64.whl (3.5MB)
Installing collected packages: confluent-kafka
Successfully installed confluent-kafka-0.11.4
You are using pip version 8.1.2, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

➜  ConfluenceKafkaClients docker run \
  --net=confluent \
confluentinc/cp-kafka:4.1.0 \
pip install --upgrade pip
Collecting pip
  Downloading https://files.pythonhosted.org/packages/0f/74/ecd13431bcc456ed390b44c8a6e917c1820365cbebcb6a8974d1cd045ab4/pip-10.0.1-py2.py3-none-any.whl (1.3MB)
Installing collected packages: pip
  Found existing installation: pip 8.1.2
    Uninstalling pip-8.1.2:
      Successfully uninstalled pip-8.1.2
Successfully installed pip-10.0.1

ConfluenceKafkaClients docker run \
  --net=confluent \
confluentinc/cp-kafka:4.1.0 \
pip freeze
You are using pip version 8.1.2, however version 10.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
backports.ssl-match-hostname==3.5.0.1
cached-property==1.4.0
confluent-docker-utils==0.0.20
docker==2.5.1
docker-compose==1.16.1
docker-pycreds==0.2.2
dockerpty==0.4.1
docopt==0.6.2
enum34==1.1.6
functools32==3.2.3.post2
ipaddress==1.0.19
Jinja2==2.9.6
jsonschema==2.6.0
MarkupSafe==1.0
PyYAML==3.12
requests==2.11.1
six==1.11.0
texttable==0.9.1
websocket-client==0.47.0

Neither the pip version was updated nor the confluent-kafka-python module was installed, maybe the command docker run ….pip install confluent-kafka, it doesn’t install the module in the container but somewhere else?

I execute the producer from mi local shared folder with the docker, and still missing the module:

➜  ConfluenceKafkaClients docker run \
  --net=confluent \
-it -v '/Volumes/G-DRIVE mobile SSD R-Series/Dockers/Users':/ConfluenceKafkaClients \
confluentinc/cp-kafka:4.1.0 \
python /ConfluenceKafkaClients/producer_example.py
Traceback (most recent call last):
  File "/ConfluenceKafkaClients/producer_example.py", line 8, in <module>
    from confluent_kafka import Producer
ImportError: No module named confluent_kafka 


Do you have any idea of what im doing wrong?

Thanks in advance for your appreciate help.

**Comment by garrisa** [comment]
I create newly, the docker-machine and the containers nodes, and run pip install confluent-kaka python client module in cp-kafka container instance:

CONTAINER ID        IMAGE                              COMMAND                  CREATED             STATUS              PORTS                          NAMES
356045bb6579        confluentinc/cp-kafka:latest       "/etc/confluent/dock…"   11 minutes ago      Up 12 minutes       9092/tcp                       kafka
7ff39220414b        confluentinc/cp-zookeeper:latest   "/etc/confluent/dock…"   14 minutes ago      Up 14 minutes       2181/tcp, 2888/tcp, 3888/tcp   zookeeper
7e70786abf6b        confluentinc/cp-kafka:latest       "/etc/confluent/dock…"   22 minutes ago      Up 22 minutes                                      kafka-single-node_kafka_1
5fb1e70690ac        confluentinc/cp-zookeeper:latest   "/etc/confluent/dock…"   22 minutes ago      Up 22 minutes                                      kafka-single-node_zookeeper_1

But I still receiving the same error, so im unable to install confluent-kaka python client module in cp-kafka container instance.

ConfluenceKafkaClients docker run \
  --net=confluent \
-it -v '/Volumes/G-DRIVE mobile SSD R-Series/Dockers/Users':/ConfluenceKafkaClients \
confluentinc/cp-kafka:latest \
> python /ConfluenceKafkaClients/producer_example.py


Traceback (most recent call last):
  File "/ConfluenceKafkaClients/producer_example.py", line 8, in <module>
    from confluent_kafka import Producer
ImportError: No module named confluent_kafka


Do you have any doc to follow the confluent_kafka install in a cp_kafka container instance?

Or run a xecution of a confluent-kafka-python producer in a Confluent Docker container for Kafka?

Thanks in advance

**Comment by garrisa** [comment]
Maybe adding the following line in the docker file of the image solves the problem:

RUN pip install --upgrade pip && pip install confluent-kafka

Can I do that?



Ivan Garrido Sanz

Email: ivan.garrido@gmail.com
Skype: ohivanivan
Mobile: +34640397215
Linkedin: https://www.linkedin.com/in/iv%C3%A1n-garrido-sanz-05aa3b33/

**Comment by garrisa** [comment]
I did 
tmp/Dockerfile:

FROM confluentinc/cp-kafka
RUN pip install --upgrade pip && pip install confluent-kafka

docker image build ./tmp -t confluentinc/cp-kafka:latest

and recreate the containers and seems to work

➜  ConfluenceKafkaClients docker run \
  --net=confluent \
-it -v '/Volumes/G-DRIVE mobile SSD R-Series/Dockers/Users':/ConfluenceKafkaClients \
confluentinc/cp-kafka:latest \
 python /ConfluenceKafkaClients/producer_example.py

%3|1527852861.051|FAIL|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused
%3|1527852861.051|ERROR|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused
%3|1527852861.051|ERROR|rdkafka#producer-1| [thrd:localhost:9092/bootstrap]: 1/1 brokers are down

➜  ConfluenceKafkaClients docker run \
  --net=confluent \
-it -v '/Volumes/G-DRIVE mobile SSD R-Series/Dockers/Users':/ConfluenceKafkaClients \
confluentinc/cp-kafka:latest \
 python /ConfluenceKafkaClients/consumer_example.py

%3|1527852938.115|FAIL|client-1#consumer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused
%3|1527852938.115|ERROR|client-1#consumer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused
%3|1527852938.115|ERROR|client-1#consumer-1| [thrd:localhost:9092/bootstrap]: 1/1 brokers are down

Them I need to know what happen with this error.

Thanks and kind regards

**Comment by szeitlin** [comment]
I am having a similar problem but only when I'm in Jupyter, which makes me suspect it has to do with librdkafka not being in my path. Normally what I do in these situations is update my bash_profile so it can find the required dependencies. In this case, I installed with homebrew and I can confirm that homebrew knows it's there if I try to reinstall, but I can't tell where it went.

**Comment by edenhill** [comment]
Each time you invoke `docker run .. ` it creates a new container.
If you need files, changes, state to persist between runs you need to build your own image based on an existing image, or make sure the docker container runs in detached mode:
https://stackoverflow.com/a/27708039/1821055

**Comment by szeitlin** [comment]
So I was able to get the producer and consumer to run, there was no problem with the docker container. It turned out the problem was with IPython and Jupyter running from the global anaconda install instead of from inside the virtual env where confluent-kafka was installed. Once I fixed that (so, to be clear, it was indeed a problem with the path), everything works (both inside and outside of Jupyter).
